File: api/semgrep.py
import json
import os
import shutil
import subprocess
import sys
import tempfile
import time
from http.server import BaseHTTPRequestHandler


LANG_EXT = {
    "python": ".py",
    "javascript": ".js",
    "typescript": ".ts",
    "java": ".java",
    "go": ".go",
    "ruby": ".rb",
    "php": ".php",
    "c": ".c",
    "cpp": ".cpp",
    "rust": ".rs",
    "kotlin": ".kt",
    "scala": ".scala",
    "swift": ".swift",
    "c#": ".cs",
    "csharp": ".cs",
}

# ── 14 套规则集，按语言动态加载 ──
# 通用规则（所有语言都跑）
COMMON_RULESETS = [
    "p/default",
    "p/owasp-top-ten",
    "p/command-injection",
    "p/jwt",
    "p/secrets",
    "p/supply-chain",
]

# 语言专属规则
LANGUAGE_RULESETS = {
    "javascript": ["p/javascript", "p/react", "p/nodejs"],
    "typescript": ["p/typescript", "p/react", "p/nodejs"],
    "python": ["p/python"],
    "go": ["p/golang"],
    "java": ["p/java"],
}

# 被排除的规则集（已知不稳定或重复）
EXCLUDED_RULESETS = {"p/ssrf", "p/r2c-security-audit", "p/sql-injection", "p/xss"}


def rulesets_for(language):
    """返回该语言应加载的规则集列表（去重，不含被排除项）"""
    lang = language.lower()
    rulesets = COMMON_RULESETS + LANGUAGE_RULESETS.get(lang, [])
    # 去重并保持顺序
    seen = set()
    result = []
    for r in rulesets:
        if r not in seen and r not in EXCLUDED_RULESETS:
            seen.add(r)
            result.append(r)
    return result


def send_json(handler, status, payload):
    data = json.dumps(payload).encode("utf-8")
    handler.send_response(status)
    handler.send_header("content-type", "application/json")
    handler.send_header("content-length", str(len(data)))
    handler.end_headers()
    handler.wfile.write(data)


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        send_json(self, 200, {"ok": True, "service": "checkaicode-semgrep"})

    def do_POST(self):
        try:
            length = int(self.headers.get("content-length", "0"))
            body = json.loads(self.rfile.read(length).decode("utf-8"))
        except Exception:
            send_json(self, 400, {"error": "Invalid JSON body"})
            return

        code = body.get("code") or ""
        language = str(body.get("language") or "javascript").lower()
        filename = str(body.get("filename") or "target")

        if not code.strip():
            send_json(self, 400, {"error": "Code required"})
            return

        ext = LANG_EXT.get(language) or os.path.splitext(filename)[1] or ".txt"
        tmp_dir = tempfile.mkdtemp(prefix="checkaicode-semgrep-")
        src_file = os.path.join(tmp_dir, "target" + ext)

        try:
            with open(src_file, "w", encoding="utf-8") as f:
                f.write(code)

            semgrep_home = os.path.join(tmp_dir, ".semgrep")
            semgrep_cache = os.path.join(tmp_dir, ".cache")
            os.makedirs(semgrep_home, exist_ok=True)
            os.makedirs(semgrep_cache, exist_ok=True)

            # ── 本地规则优先，远端 fallback ──
            rulesets = rulesets_for(language)
            local_rules_dir = os.path.join(os.path.dirname(__file__), "..", "semgrep-rules")
            final_configs = []
            for rs in rulesets:
                local_yaml = os.path.join(local_rules_dir, rs.replace("/", "_") + ".yaml")
                if os.path.exists(local_yaml):
                    final_configs.append(local_yaml)
                else:
                    final_configs.append(rs)

            cmd = [
                sys.executable,
                "-c",
                "from semgrep.console_scripts.pysemgrep import main; main()",
                "scan",
                "--json",
                "--quiet",
            ]

            for cfg in final_configs:
                cmd.extend(["--config", cfg])

            cmd.append(src_file)
            env = {
                **os.environ,
                "HOME": tmp_dir,
                "XDG_CONFIG_HOME": tmp_dir,
                "XDG_CACHE_HOME": semgrep_cache,
                "SEMGREP_SETTINGS_FILE": os.path.join(semgrep_home, "settings.yml"),
                "SEMGREP_LOG_FILE": os.path.join(semgrep_home, "semgrep.log"),
                "SEMGREP_ENABLE_VERSION_CHECK": "0",
                "SEMGREP_SEND_METRICS": "off",
                "PYTHONPATH": os.pathsep.join(
                    [
                        "/var/task/_vendor",
                        "/var/task",
                        os.environ.get("PYTHONPATH", ""),
                    ]
                ),
            }

            started = time.time()
            proc = subprocess.run(
                cmd,
                cwd=tmp_dir,
                env=env,
                text=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                timeout=55,
                check=False,
            )
            scan_time_ms = int((time.time() - started) * 1000)

            try:
                output = json.loads(proc.stdout or "{}")
            except json.JSONDecodeError:
                output = {"results": [], "errors": []}

            for result in output.get("results", []):
                if isinstance(result.get("path"), str):
                    result["path"] = result["path"].replace(tmp_dir, "input")

            send_json(
                self,
                200,
                {
                    "results": output.get("results", []),
                    "errors": output.get("errors", []),
                    "scanTimeMs": scan_time_ms,
                    "exitCode": proc.returncode,
                    "stderr": proc.stderr[-4000:] if proc.returncode != 0 and not output.get("results") else "",
                    "rulesets": rulesets,  # 返回实际使用的规则集，便于调试
                },
            )
        except subprocess.TimeoutExpired:
            send_json(
                self,
                200,
                {
                    "results": [],
                    "errors": [{"message": "Semgrep scan timed out"}],
                    "scanTimeMs": 55000,
                },
            )
        except Exception as exc:
            send_json(
                self,
                200,
                {
                    "results": [],
                    "errors": [{"message": str(exc)}],
                    "scanTimeMs": 0,
                },
            )
        finally:
            shutil.rmtree(tmp_dir, ignore_errors=True)

File: src/lib/semgrep.ts
import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

interface SemgrepFinding {
  check_id: string;
  path: string;
  start: { line: number; col: number };
  end: { line: number; col: number };
  extra: {
    message: string;
    metadata?: {
      cwe?: string[];
      owasp?: string[];
      severity?: string;
      confidence?: string;
    };
    severity?: string;
    lines?: string;
    fix?: string;
  };
}

interface SemgrepOutput {
  results: SemgrepFinding[];
  errors: any[];
}

interface SemgrepApiOutput {
  results: SemgrepFinding[];
  errors: any[];
  scanTimeMs: number;
  exitCode?: number;
  stderr?: string;
}

const LANG_EXT: Record<string, string> = {
  python: ".py",
  javascript: ".js",
  typescript: ".ts",
  java: ".java",
  go: ".go",
  ruby: ".rb",
  php: ".php",
  c: ".c",
  cpp: ".cpp",
  rust: ".rs",
  kotlin: ".kt",
  scala: ".scala",
  swift: ".swift",
  "c#": ".cs",
  csharp: ".cs",
};

/**
 * 从代码内容或文件名推断语言
 */
function detectLanguage(code: string, filename?: string): string {
  // 优先从文件名推断
  if (filename) {
    const ext = filename.split(".").pop()?.toLowerCase();
    if (ext) {
      const langByExt = Object.entries(LANG_EXT).find(([, e]) => e === `.${ext}`);
      if (langByExt) return langByExt[0];
    }
  }
  // 从代码内容推断
  if (code.includes("def ") || code.includes("import ") || code.includes("print(")) return "python";
  if (code.includes("package main") || code.includes("func ")) return "go";
  if (code.includes("public class ") || code.includes("import java.")) return "java";
  if (code.includes("interface ") && code.includes(": ")) return "typescript";
  if (code.includes("const ") || code.includes("let ") || code.includes("var ")) return "javascript";
  if (code.includes("function ") || code.includes("=>")) return "javascript";
  return "javascript"; // 默认
}

function mapSemgrepSeverity(s: string | undefined): "critical" | "warning" | "info" {
  if (!s) return "warning";
  const sl = s.toLowerCase();
  if (sl === "error" || sl === "critical") return "critical";
  if (sl === "warning") return "warning";
  return "info";
}

function mapCheckIdToType(checkId: string): "security" | "quality" | "dependency" {
  const id = checkId.toLowerCase();
  if (
    id.includes("sql") ||
    id.includes("inject") ||
    id.includes("xss") ||
    id.includes("command") ||
    id.includes("eval") ||
    id.includes("exec") ||
    id.includes("secret") ||
    id.includes("key") ||
    id.includes("password") ||
    id.includes("csrf") ||
    id.includes("ssrf") ||
    id.includes("traversal") ||
    id.includes("deserialization") ||
    id.includes("crypto") ||
    id.includes("hash") ||
    id.includes("cipher") ||
    id.includes("cors") ||
    id.includes("debug")
  ) {
    return "security";
  }
  return "quality";
}

function cweReferenceUrl(cwe: string | undefined): string | undefined {
  const match = cwe?.match(/CWE-(\d+)/i);
  return match ? `https://cwe.mitre.org/data/definitions/${match[1]}.html` : undefined;
}

function parseSemgrepOutput(output: SemgrepApiOutput, tmpDir: string) {
  const issues = output.results.map((r) => {
    const severity = mapSemgrepSeverity(
      r.extra?.severity || r.extra?.metadata?.severity
    );
    const type = mapCheckIdToType(r.check_id);
    const fix = r.extra?.fix;

    return {
      type,
      severity,
      file: r.path.replace(tmpDir, "input"),
      line: r.start.line,
      endLine: r.end.line,
      message: r.extra.message,
      ruleId: r.check_id,
      fixSuggestion: fix
        ? `Suggested fix: ${fix.trim().slice(0, 200)}`
        : undefined,
      fixCode: fix || undefined,
      codeSnippet: r.extra?.lines?.trim() || undefined,
      referenceUrl: cweReferenceUrl(r.extra?.metadata?.cwe?.[0]),
    };
  });

  return {
    issues,
    scanTimeMs: output.scanTimeMs,
    error: output.errors?.length
      ? output.errors.map((e: any) => e.message || String(e)).join("; ")
      : undefined,
  };
}

export async function runSemgrep(code: string, language: string): Promise<{
  issues: {
    type: "security" | "quality" | "dependency";
    severity: "critical" | "warning" | "info";
    file: string;
    line: number;
    endLine: number;
    message: string;
    ruleId: string;
    fixSuggestion?: string;
    fixCode?: string;
    codeSnippet?: string;
    referenceUrl?: string;
  }[];
  scanTimeMs: number;
  error?: string;
}> {
  // Local development may opt in to a machine-installed Semgrep CLI.
  // Production defaults to the Python Function because the Node runtime
  // cannot reliably execute the Python Semgrep package.
  if (process.env.SEMGREP_USE_LOCAL === "1") {
    try {
      execSync("which semgrep", { encoding: "utf-8", stdio: "pipe" });
      return runSemgrepLocal(code, language);
    } catch {
      // CLI not available, use HTTP API
    }
  }

  const baseUrl =
    process.env.SEMGREP_API_URL ||
    process.env.NEXTAUTH_URL ||
    (process.env.VERCEL ? "https://checkaicode.com" : "http://localhost:3000");

  const headers: Record<string, string> = {
    "content-type": "application/json",
  };

  if (process.env.SEMGREP_API_SECRET) {
    headers["x-semgrep-api-secret"] = process.env.SEMGREP_API_SECRET;
  }

  const detectedLang = language === "auto" || !language
    ? detectLanguage(code)
    : language;

  const res = await fetch(`${baseUrl}/api/semgrep`, {
    method: "POST",
    headers,
    body: JSON.stringify({
      code,
      language: detectedLang,
      filename: "input",
    }),
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "Unknown error");
    return { issues: [], scanTimeMs: 0, error: `Semgrep API error ${res.status}: ${text}` };
  }

  const output: SemgrepApiOutput = await res.json();

  // Use a dummy tmpDir for path replacement (Python Function already replaced tmp_dir with "input")
  return parseSemgrepOutput(output, "input");
}

// ── 14 套规则集，按语言动态加载 ──
const COMMON_RULESETS = [
  "p/default",
  "p/owasp-top-ten",
  "p/command-injection",
  "p/jwt",
  "p/secrets",
  "p/supply-chain",
];

const LANGUAGE_RULESETS: Record<string, string[]> = {
  javascript: ["p/javascript", "p/react", "p/nodejs"],
  typescript: ["p/typescript", "p/react", "p/nodejs"],
  python: ["p/python"],
  go: ["p/golang"],
  java: ["p/java"],
};

const EXCLUDED_RULESETS = new Set([
  "p/ssrf",
  "p/r2c-security-audit",
  "p/sql-injection",
  "p/xss",
]);

function rulesetsFor(language: string): string[] {
  const lang = language.toLowerCase();
  const rulesets = [...COMMON_RULESETS, ...(LANGUAGE_RULESETS[lang] || [])];
  const seen = new Set<string>();
  const result: string[] = [];
  for (const r of rulesets) {
    if (!seen.has(r) && !EXCLUDED_RULESETS.has(r)) {
      seen.add(r);
      result.push(r);
    }
  }
  return result;
}

async function runSemgrepLocal(code: string, language: string): Promise<{
  issues: {
    type: "security" | "quality" | "dependency";
    severity: "critical" | "warning" | "info";
    file: string;
    line: number;
    endLine: number;
    message: string;
    ruleId: string;
    fixSuggestion?: string;
    fixCode?: string;
    codeSnippet?: string;
    referenceUrl?: string;
  }[];
  scanTimeMs: number;
  error?: string;
}> {
  const ext = LANG_EXT[language.toLowerCase()] || ".txt";
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "checkaicode-semgrep-"));
  const srcFile = path.join(tmpDir, `target${ext}`);

  const activeRulesets = rulesetsFor(language);

  try {
    fs.writeFileSync(srcFile, code, "utf-8");

    const startTime = Date.now();
    const cmdParts = ["semgrep", "--json", "--quiet"];
    for (const rs of activeRulesets) {
      cmdParts.push("--config", rs);
    }
    cmdParts.push(srcFile);

    const stdout = execSync(cmdParts.join(" "), {
      encoding: "utf-8",
      timeout: 30000,
      maxBuffer: 10 * 1024 * 1024,
    });

    const scanTimeMs = Date.now() - startTime;
    const output: SemgrepOutput = JSON.parse(stdout);

    const issues = output.results.map((r) => {
      const severity = mapSemgrepSeverity(
        r.extra?.severity || r.extra?.metadata?.severity
      );
      const type = mapCheckIdToType(r.check_id);
      const fix = r.extra?.fix;

      return {
        type,
        severity,
        file: r.path.replace(tmpDir, "input"),
        line: r.start.line,
        endLine: r.end.line,
        message: r.extra.message,
        ruleId: r.check_id,
        fixSuggestion: fix
          ? `Suggested fix: ${fix.trim().slice(0, 200)}`
          : undefined,
        fixCode: fix || undefined,
        codeSnippet: r.extra?.lines?.trim() || undefined,
        referenceUrl: cweReferenceUrl(r.extra?.metadata?.cwe?.[0]),
      };
    });

    return { issues, scanTimeMs };
  } catch (err: any) {
    if (err.status === 1 && err.stdout) {
      try {
        const output: SemgrepOutput = JSON.parse(err.stdout);
        const scanTimeMs = 0;
        const issues = output.results.map((r) => {
          const severity = mapSemgrepSeverity(
            r.extra?.severity || r.extra?.metadata?.severity
          );
          const type = mapCheckIdToType(r.check_id);
          const fix = r.extra?.fix;
          return {
            type,
            severity,
            file: r.path.replace(tmpDir, "input"),
            line: r.start.line,
            endLine: r.end.line,
            message: r.extra.message,
            ruleId: r.check_id,
            fixSuggestion: fix
              ? `Suggested fix: ${fix.trim().slice(0, 200)}`
              : undefined,
            fixCode: fix || undefined,
            codeSnippet: r.extra?.lines?.trim() || undefined,
            referenceUrl: cweReferenceUrl(r.extra?.metadata?.cwe?.[0]),
          };
        });
        return { issues, scanTimeMs };
      } catch {
        return { issues: [], scanTimeMs: 0, error: err.message };
      }
    }
    return { issues: [], scanTimeMs: 0, error: err.message };
  } finally {
    try {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    } catch {
      // ignore cleanup errors
    }
  }
}
